File: src/services/organization_service.py
"""
Organization service for KindnessHome platform
Business logic layer for managing charitable organization data
"""

import logging
from typing import Optional, List, Dict, Any
from datetime import datetime, timedelta

# Import our custom modules (these will be added to the backend)
from .charity_api_service import CharityAPIService
from ..models.organization import Organization, OrganizationCache

logger = logging.getLogger(__name__)

class OrganizationService:
    """Service for managing charitable organization data"""
    
    def __init__(self, charity_api_key: str = None):
        self.charity_api = CharityAPIService(charity_api_key)
        self.cache = OrganizationCache()
        
    def search_organizations(self, 
                           query: str, 
                           state: str = None, 
                           city: str = None,
                           category: str = None,
                           limit: int = 20,
                           use_cache: bool = True) -> List[Organization]:
        """
        Search for organizations with caching support
        
        Args:
            query: Search term for organization name
            state: Filter by state code (optional)
            city: Filter by city (optional) 
            category: Filter by NTEE category (optional)
            limit: Maximum number of results
            use_cache: Whether to check cache first
            
        Returns:
            List of Organization objects
        """
        try:
            organizations = []
            
            # First, try cache if enabled
            if use_cache:
                cached_results = self.cache.search_organizations(query, state, limit)
                if cached_results:
                    logger.info(f"Found {len(cached_results)} cached results for: {query}")
                    organizations.extend(cached_results)
                    
                    # If we have enough cached results, return them
                    if len(organizations) >= limit:
                        return organizations[:limit]
            
            # Search CharityAPI for additional results
            try:
                api_results = self.charity_api.search_organizations(
                    query=query,
                    city=city,
                    state=state,
                    limit=limit - len(organizations)
                )
                
                for org_data in api_results:
                    org = Organization.from_charity_api(org_data)
                    
                    # Filter by category if specified
                    if category and org.ntee_code:
                        if not org.ntee_code.upper().startswith(category.upper()):
                            continue
                    
                    organizations.append(org)
                    
                    # Cache the organization for future searches
                    self.cache.save_organization(org)
                    
                logger.info(f"Found {len(api_results)} API results for: {query}")
                
            except Exception as e:
                logger.error(f"CharityAPI search failed: {str(e)}")
                # Continue with cached results if API fails
            
            # Remove duplicates (by EIN) and limit results
            seen_eins = set()
            unique_orgs = []
            
            for org in organizations:
                if org.ein not in seen_eins:
                    seen_eins.add(org.ein)
                    unique_orgs.append(org)
                    
                if len(unique_orgs) >= limit:
                    break
            
            return unique_orgs
            
        except Exception as e:
            logger.error(f"Error in organization search: {str(e)}")
            return []

    def get_organization_by_ein(self, ein: str, use_cache: bool = True) -> Optional[Organization]:
        """
        Get organization by EIN with caching support
        
        Args:
            ein: Employer Identification Number
            use_cache: Whether to check cache first
            
        Returns:
            Organization object or None if not found
        """
        try:
            # Clean EIN
            clean_ein = ein.replace('-', '').replace(' ', '')
            
            # Check cache first if enabled
            if use_cache:
                cached_org = self.cache.get_organization(clean_ein)
                if cached_org:
                    # Check if cache is still fresh (30 days)
                    if cached_org.last_updated and \
                       (datetime.utcnow() - cached_org.last_updated).days < 30:
                        logger.info(f"Found cached organization: {cached_org.name}")
                        self.cache.increment_search_count(clean_ein)
                        return cached_org
            
            # Fetch from CharityAPI
            try:
                org_data = self.charity_api.get_organization_by_ein(clean_ein)
                
                if org_data:
                    org = Organization.from_charity_api(org_data)
                    
                    # Cache the organization
                    self.cache.save_organization(org)
                    self.cache.increment_search_count(clean_ein)
                    
                    logger.info(f"Found organization via API: {org.name}")
                    return org
                else:
                    logger.info(f"Organization not found for EIN: {clean_ein}")
                    return None
                    
            except Exception as e:
                logger.error(f"CharityAPI lookup failed: {str(e)}")
                
                # Return cached version even if stale, if available
                if use_cache:
                    cached_org = self.cache.get_organization(clean_ein)
                    if cached_org:
                        logger.warning(f"Returning stale cached data for EIN: {clean_ein}")
                        return cached_org
                
                return None
                
        except Exception as e:
            logger.error(f"Error getting organization by EIN: {str(e)}")
            return None

    def verify_organization(self, ein: str) -> Dict[str, Any]:
        """
        Verify organization's charitable status and tax-deductible eligibility
        
        Args:
            ein: Employer Identification Number
            
        Returns:
            Dictionary with verification results
        """
        try:
            org = self.get_organization_by_ein(ein)
            
            if not org:
                return {
                    'verified': False,
                    'exists': False,
                    'message': 'Organization not found in IRS database'
                }
            
            verification = {
                'verified': True,
                'exists': True,
                'ein': org.ein,
                'name': org.name,
                'is_public_charity': org.is_public_charity(),
                'is_tax_deductible': org.is_tax_deductible(),
                'status': org.status,
                'classification': org.classification,
                'deductibility': org.deductibility
            }
            
            # Add status message
            if org.is_public_charity():
                verification['message'] = 'Verified 501(c)(3) public charity - donations are tax-deductible'
            elif org.is_tax_deductible():
                verification['message'] = 'Tax-exempt organization - donations may be tax-deductible'
            else:
                verification['message'] = 'Tax-exempt organization - donations may not be tax-deductible'
            
            return verification
            
        except Exception as e:
            logger.error(f"Error verifying organization: {str(e)}")
            return {
                'verified': False,
                'exists': False,
                'message': 'Error verifying organization status'
            }

    def get_popular_organizations(self, limit: int = 20) -> List[Organization]:
        """
        Get most popular/searched organizations
        
        Args:
            limit: Maximum number of results
            
        Returns:
            List of popular Organization objects
        """
        try:
            return self.cache.get_popular_organizations(limit)
        except Exception as e:
            logger.error(f"Error getting popular organizations: {str(e)}")
            return []

    def get_organizations_by_category(self, category: str, limit: int = 20) -> List[Organization]:
        """
        Get organizations by NTEE category
        
        Args:
            category: NTEE major group code (A-Z)
            limit: Maximum number of results
            
        Returns:
            List of Organization objects in the category
        """
        try:
            # This is a simplified implementation
            # In a full implementation, you might want to pre-populate
            # popular organizations by category
            
            # For now, we'll search for organizations with the category code
            # This could be enhanced with category-specific searches
            
            category_names = {
                'A': 'arts culture humanities',
                'B': 'education school university',
                'C': 'environment animals wildlife',
                'D': 'animal welfare pets',
                'E': 'health medical hospital',
                'F': 'mental health crisis',
                'G': 'disease medical research',
                'H': 'medical research',
                'I': 'crime legal justice',
                'J': 'employment job training',
                'K': 'food nutrition hunger',
                'L': 'housing shelter homeless',
                'M': 'disaster relief emergency',
                'N': 'recreation sports athletics',
                'O': 'youth development children',
                'P': 'human services social',
                'Q': 'international foreign',
                'R': 'civil rights advocacy',
                'S': 'community development',
                'T': 'philanthropy foundation',
                'U': 'science technology research',
                'V': 'social science research',
                'W': 'public benefit society',
                'X': 'religion religious',
                'Y': 'membership benefit',
                'Z': 'unknown'
            }
            
            search_term = category_names.get(category.upper(), 'charity')
            return self.search_organizations(search_term, category=category, limit=limit)
            
        except Exception as e:
            logger.error(f"Error getting organizations by category: {str(e)}")
            return []

    def get_categories(self) -> List[Dict[str, Any]]:
        """
        Get list of available organization categories (NTEE codes)
        
        Returns:
            List of category dictionaries
        """
        from .charity_api_service import get_ntee_categories
        
        categories = []
        ntee_map = get_ntee_categories()
        
        for code, name in ntee_map.items():
            categories.append({
                'code': code,
                'name': name,
                'description': f"Organizations focused on {name.lower()}"
            })
        
        return categories

    def sync_popular_organizations(self, limit: int = 100):
        """
        Background task to sync popular organizations from CharityAPI
        This could be run periodically to keep cache fresh
        
        Args:
            limit: Number of organizations to sync
        """
        try:
            logger.info("Starting popular organizations sync...")
            
            # Get a list of well-known charity EINs to pre-populate
            # This is a simplified approach - in production you might
            # use a curated list or analyze search patterns
            
            well_known_eins = [
                "530196605",  # American Red Cross
                "131624016",  # Salvation Army
                "541760264",  # Doctors Without Borders
                "237327031",  # United Way
                "521382983",  # Habitat for Humanity
                "363271564",  # St. Jude Children's Research Hospital
                "237327031",  # Goodwill Industries
                "131740204",  # YMCA
                "237327031",  # Boys & Girls Clubs
                "521382983"   # Feeding America
            ]
            
            synced_count = 0
            
            for ein in well_known_eins:
                try:
                    org = self.get_organization_by_ein(ein, use_cache=False)
                    if org:
                        synced_count += 1
                        logger.info(f"Synced: {org.name}")
                        
                except Exception as e:
                    logger.error(f"Error syncing EIN {ein}: {str(e)}")
                    continue
                    
                if synced_count >= limit:
                    break
            
            logger.info(f"Completed sync of {synced_count} popular organizations")
            
        except Exception as e:
            logger.error(f"Error in popular organizations sync: {str(e)}")

    def test_api_connection(self) -> Dict[str, Any]:
        """
        Test the CharityAPI connection and return status
        
        Returns:
            Dictionary with connection test results
        """
        try:
            is_connected = self.charity_api.test_connection()
            rate_limit_info = self.charity_api.get_rate_limit_info()
            
            return {
                'connected': is_connected,
                'api_key_valid': bool(self.charity_api.api_key),
                'rate_limit': rate_limit_info,
                'timestamp': datetime.utcnow().isoformat()
            }
            
        except Exception as e:
            logger.error(f"Error testing API connection: {str(e)}")
            return {
                'connected': False,
                'api_key_valid': False,
                'error': str(e),
                'timestamp': datetime.utcnow().isoformat()
            }

File: src/services/organizationService.js
/**
 * Organization API Service for KindnessHome Frontend
 * Handles all API calls to the backend for organization data
 */

const API_BASE_URL = 'https://kindnesshome-backend.onrender.com/api';

class OrganizationService {
  /**
   * Search for organizations by query, location, and category
   * @param {Object} params - Search parameters
   * @param {string} params.query - Search term for organization name
   * @param {string} params.state - State filter (e.g., 'CA', 'NY')
   * @param {string} params.city - City filter
   * @param {string} params.category - NTEE category code (A-Z)
   * @param {number} params.limit - Maximum number of results (default 20)
   * @returns {Promise<Object>} Search results with organizations array
   */
  async searchOrganizations({ query, state, city, category, limit = 20 }) {
    try {
      const params = new URLSearchParams();
      
      if (query) params.append('q', query);
      if (state) params.append('state', state.toUpperCase());
      if (city) params.append('city', city);
      if (category) params.append('category', category.toUpperCase());
      if (limit) params.append('limit', limit.toString());

      const response = await fetch(`${API_BASE_URL}/organizations/search?${params}`);
      
      if (!response.ok) {
        throw new Error(`Search failed: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      return {
        organizations: data.data || [],
        count: data.count || 0,
        query: data.query || '',
        filters: data.filters || {}
      };
    } catch (error) {
      console.error('Error searching organizations:', error);
      throw new Error('Failed to search organizations. Please try again.');
    }
  }

  /**
   * Get organization details by EIN (Tax ID)
   * @param {string} ein - Employer Identification Number
   * @returns {Promise<Object>} Organization details
   */
  async getOrganizationByEin(ein) {
    try {
      const cleanEin = ein.replace(/[-\s]/g, '');
      const response = await fetch(`${API_BASE_URL}/organizations/${cleanEin}`);
      
      if (!response.ok) {
        if (response.status === 404) {
          throw new Error('Organization not found');
        }
        throw new Error(`Failed to get organization: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      return data.data;
    } catch (error) {
      console.error('Error getting organization:', error);
      throw error;
    }
  }

  /**
   * Verify organization's charitable status and tax-deductible eligibility
   * @param {string} ein - Employer Identification Number
   * @returns {Promise<Object>} Verification results
   */
  async verifyOrganization(ein) {
    try {
      const cleanEin = ein.replace(/[-\s]/g, '');
      const response = await fetch(`${API_BASE_URL}/organizations/verify/${cleanEin}`);
      
      if (!response.ok) {
        throw new Error(`Verification failed: ${response.status} ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error verifying organization:', error);
      throw new Error('Failed to verify organization status.');
    }
  }

  /**
   * Get popular/most searched organizations
   * @param {number} limit - Maximum number of results (default 20)
   * @returns {Promise<Array>} Array of popular organizations
   */
  async getPopularOrganizations(limit = 20) {
    try {
      const response = await fetch(`${API_BASE_URL}/organizations/popular?limit=${limit}`);
      
      if (!response.ok) {
        throw new Error(`Failed to get popular organizations: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      return data.data || [];
    } catch (error) {
      console.error('Error getting popular organizations:', error);
      throw new Error('Failed to load popular organizations.');
    }
  }

  /**
   * Get all available organization categories (NTEE codes)
   * @returns {Promise<Array>} Array of category objects
   */
  async getCategories() {
    try {
      const response = await fetch(`${API_BASE_URL}/organizations/categories`);
      
      if (!response.ok) {
        throw new Error(`Failed to get categories: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      return data.data || [];
    } catch (error) {
      console.error('Error getting categories:', error);
      throw new Error('Failed to load categories.');
    }
  }

  /**
   * Get organizations by specific category
   * @param {string} categoryCode - NTEE category code (A-Z)
   * @param {number} limit - Maximum number of results (default 20)
   * @returns {Promise<Array>} Array of organizations in the category
   */
  async getOrganizationsByCategory(categoryCode, limit = 20) {
    try {
      const response = await fetch(`${API_BASE_URL}/organizations/categories/${categoryCode.toUpperCase()}?limit=${limit}`);
      
      if (!response.ok) {
        throw new Error(`Failed to get organizations by category: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      return {
        organizations: data.data || [],
        count: data.count || 0,
        category: data.category || categoryCode
      };
    } catch (error) {
      console.error('Error getting organizations by category:', error);
      throw new Error('Failed to load organizations for this category.');
    }
  }

  /**
   * Test API connection and get status
   * @returns {Promise<Object>} Connection test results
   */
  async testConnection() {
    try {
      const response = await fetch(`${API_BASE_URL}/organizations/test`);
      
      if (!response.ok) {
        throw new Error(`Connection test failed: ${response.status} ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error testing API connection:', error);
      throw new Error('Failed to connect to organization database.');
    }
  }

  /**
   * Format organization data for display
   * @param {Object} org - Raw organization data from API
   * @returns {Object} Formatted organization data
   */
  formatOrganization(org) {
    return {
      ...org,
      // Format EIN with dashes for display
      formattedEin: org.ein ? `${org.ein.slice(0, 2)}-${org.ein.slice(2)}` : '',
      
      // Create full address string
      fullAddress: [org.street, org.city, org.state, org.zip_code]
        .filter(Boolean)
        .join(', '),
      
      // Format revenue amount
      formattedRevenue: this.formatCurrency(org.revenue_amount),
      
      // Format asset amount
      formattedAssets: this.formatCurrency(org.asset_amount),
      
      // Create verification badges
      badges: {
        verified: org.is_public_charity && org.is_tax_deductible,
        taxDeductible: org.is_tax_deductible,
        publicCharity: org.is_public_charity
      }
    };
  }

  /**
   * Format currency amounts for display
   * @param {number} amount - Amount in dollars
   * @returns {string} Formatted currency string
   */
  formatCurrency(amount) {
    if (!amount || amount === 0) return 'Not Available';
    
    if (amount >= 1000000000) {
      return `$${(amount / 1000000000).toFixed(1)}B`;
    } else if (amount >= 1000000) {
      return `$${(amount / 1000000).toFixed(1)}M`;
    } else if (amount >= 1000) {
      return `$${(amount / 1000).toFixed(1)}K`;
    } else {
      return `$${amount.toLocaleString()}`;
    }
  }

  /**
   * Get organization category color for UI
   * @param {string} categoryCode - NTEE category code
   * @returns {string} CSS color class or hex color
   */
  getCategoryColor(categoryCode) {
    const colors = {
      'A': '#FF6B6B', // Arts - Red
      'B': '#4ECDC4', // Education - Teal
      'C': '#45B7D1', // Environment - Blue
      'D': '#96CEB4', // Animals - Green
      'E': '#FFEAA7', // Health - Yellow
      'F': '#DDA0DD', // Mental Health - Purple
      'G': '#FFB347', // Diseases - Orange
      'H': '#98D8C8', // Medical Research - Mint
      'I': '#F7DC6F', // Crime Prevention - Light Yellow
      'J': '#BB8FCE', // Employment - Light Purple
      'K': '#85C1E9', // Food/Agriculture - Light Blue
      'L': '#F8C471', // Housing - Peach
      'M': '#82E0AA', // Public Safety - Light Green
      'N': '#F1948A', // Recreation - Pink
      'O': '#D7BDE2', // Religion - Lavender
      'P': '#AED6F1', // Human Services - Sky Blue
      'Q': '#A9DFBF', // International - Sage
      'R': '#F9E79F', // Civil Rights - Cream
      'S': '#FADBD8', // Community Improvement - Rose
      'T': '#D5DBDB', // Philanthropy - Gray
      'U': '#FCF3CF', // Science - Light Cream
      'V': '#EBDEF0', // Social Science - Light Lavender
      'W': '#E8F8F5', // Public Benefit - Mint Cream
      'X': '#FEF9E7', // Religion Related - Ivory
      'Y': '#EAEDED', // Mutual Benefit - Light Gray
      'Z': '#F4F6F6'  // Unknown - Very Light Gray
    };
    
    return colors[categoryCode?.toUpperCase()] || colors['Z'];
  }
}

// Create and export a singleton instance
const organizationService = new OrganizationService();
export default organizationService;

// Also export the class for testing
export { OrganizationService };
